lib/StackAnalysis/SegregateDirectStackAccesses.cpp

```diff
@@ -0,0 +1,152 @@
+  Context = &(F.getContext());
+
+  // Get the result of the GCBI analysis
+  auto &M = *F.getParent();
+  const auto &MAMProxy = FAM.getResult<ModuleAnalysisManagerFunctionProxy>(F);
+  GCBI = MAMProxy.getCachedResult<GeneratedCodeBasicInfoAnalysis>(M);
+  if (!GCBI)
+    GCBI = &(FAM.getResult<GeneratedCodeBasicInfoAnalysis>(F));
+  revng_assert(GCBI != nullptr);
+
+  // Populate the two buckets with all load and store instruction of the
+  // function, properly segregated.
+  segregateAccesses(F);
+
+  // Adorn IR with the alias information collected before.
+  decorateStackAccesses();
+
+  return PreservedAnalyses::none();
+}
+
+void SDSAP::segregateAccesses(Function &F) {
+  using namespace PatternMatch;
+  Value *LoadSP = nullptr;
+  bool Found = false;
+
+  for (BasicBlock &BB : F) {
+    for (Instruction &I : BB) {
+      if (auto *LI = dyn_cast<LoadInst>(&I)) {
+        if (GCBI->isSPReg(skipCasts(LI->getPointerOperand()))) {
+          LoadSP = LI;
+          Found = true;
+          break;
+        }
+      }
+    }
+
+    if (Found)
+      break;
+  }
+
+  // Context: inttoptr instructions basically inhibits all optimizations. In
+  // particular, when an integer is inttoptr'd twice with different destination
+  // type, alias analysis messes up. Hence, we need to ensure that no inttoptr
+  // exists when operating on a instruction that directly accesses the stack.
+  // Note that this problem will be addressed by opaque pointers in the future.
+  IRBuilder<> Builder(&F.getEntryBlock().front());
+  auto *I8PtrTy = Builder.getInt8PtrTy();
+  Value *SPI8Ptr = Builder
+                     .CreateLoad(I8PtrTy,
+                                 ConstantExpr::getBitCast(GCBI->spReg(),
+                                                          I8PtrTy
+                                                            ->getPointerTo()));
+
+  for (BasicBlock &BB : F) {
+    for (Instruction &I : BB) {
+      Value *Ptr = nullptr;
+
+      // Differentiate accesses and add them onto their respective bucket.
+      // Everything that is not a direct access on the stack is put onto the
+      // bucket `NotDirectStackAccesses`. Load/store that access the CSVs will
+      // have their alias info added later as well.
+      if (auto *LI = dyn_cast<LoadInst>(&I))
+        Ptr = skipCasts(LI->getPointerOperand());
+      else if (auto *SI = dyn_cast<StoreInst>(&I))
+        Ptr = skipCasts(SI->getPointerOperand());
+
+      if (Ptr) {
+        Value *LoadPtr, *LHS, *BitCast;
+        ConstantInt *Offset;
+        Builder.SetInsertPoint(&I);
+
+        // Do we have a inttoptr to a load i64, i64* LoadPtr as pointer operand
+        // of the current instruction, where LoadPtr is SP? Change it with the
+        // newly-created bitcasted load in order to prevent from using inttoptr.
+        if (match(Ptr, m_Load(m_Value(LoadPtr)))
+            && GCBI->isSPReg(skipCasts(LoadPtr))) {
```

Comment:
Below you use `LoadSP`. Why not be coherent?
Checking `isSPReg` seems safer since it does not assume we a have a single `load sp`.

Reply:
I'm not sure, so you're saying to change
``` cpp
else if (match(Pointer, m_c_Add(m_Value(LHS), m_ConstantInt(Offset)))
                   && LHS == LoadSP)
```
into
``` cpp
else if (match(Pointer, m_c_Add(m_Value(LHS), m_ConstantInt(Offset)))
                   && GCBI->isSPReg(skipCasts(LHS))
```
? I need to check the loaded value, not SP.

Reply:
OK, then, viceversa, would the following work here?

```suggestion
            && LoadPtr == LoadSP)) {
```

Reply:
I got confused, I was looking at the first of the two `if`s.
What I'm saying is: if you're sure (since you asserted) there's only a load from `sp`, don't use `GCBI->isSPReg()` but just compare against `LoadSP` everywhere.

`skipCasts(LoadPtr) == LoadSP` sounds good.

(GitHub is not showing me the full diff :( right now)